Allow `:search`ing partially-applied SAW context types

samcowger opened the pull request:
As in, allow a user to write `:search ProofScript` at the SAW REPL, rather than requiring them to write `:search (ProofScript _)` - though the latter is still allowed.

sauclovian-g commented:
There are, I'm afraid, some holes in this as it stands; e.g. `:search {a} ProofScript a` generates a parse error, although `:search {a} (ProofScript a)` still works. Also, the idea of the list of `BaseType` in `SchemaPattern` is that you can do `:search Int String`, and with these changes, while you can still do `:search (ProofScript _) Int String`, both `:search ProofScript Int String` and `:search Int String ProofScript` fail.

I think the right thing to do is fold `Context` into `BaseType`; there's no reason to distinguish them in the parser, and there doesn't seem to be any problem with having sequences of `BaseType` in `AppliedType`. This allows one to write uninhabitable types like `Int Int` but these days the typechecker should cope.

(also, while you're touching this file, could you delete the `CryptolSetup` keyword on line 67? it is not used and generates a warning)

sauclovian-g commented:
Also it might be a good idea to replace the note in the docs about needing `(ProofScript _)` with one about how `:search ProofScript Int` will be interpreted as multiple search terms rather than a type application, and to use `:search (ProofScript Int)` if you want the type application.

samcowger commented:
> There are, I'm afraid, some holes in this as it stands; e.g. `:search {a} ProofScript a` generates a parse error, although `:search {a} (ProofScript a)` still works. Also, the idea of the list of `BaseType` in `SchemaPattern` is that you can do `:search Int String`, and with these changes, while you can still do `:search (ProofScript _) Int String`, both `:search ProofScript Int String` and `:search Int String ProofScript` fail.
>
> I think the right thing to do is fold `Context` into `BaseType`; there's no reason to distinguish them in the parser, and there doesn't seem to be any problem with having sequences of `BaseType` in `AppliedType`. This allows one to write uninhabitable types like `Int Int` but these days the typechecker should cope.

I'm pleasantly surprised to see that the parser doesn't complain about that change, so I've made it in d4c37a46e6fbe2a8acdd16ade2efbed02d119363. I also added tests in that commit for those now-no-longer-failing cases you discovered - thanks for discovering them!

> (also, while you're touching this file, could you delete the `CryptolSetup` keyword on line 67? it is not used and generates a warning)

Ah, thanks for the reminder, I noticed that as well. I've removed that, as well as one other non-user-facing [occurrence](https://github.com/GaloisInc/saw-script/blob/0bd979d0f7b0c5022cfde35085369b66f77d7af0/doc/saw-lexer/src/saw_lexer.py#L44) of `CryptolSetup`, in 99412c35a2e13c019093ffd9f2a8ea6ca5fb5768.

> Also it might be a good idea to replace the note in the docs about needing `(ProofScript _)` with one about how `:search ProofScript Int` will be interpreted as multiple search terms rather than a type application, and to use `:search (ProofScript Int)` if you want the type application.

Good idea, done in 494810438d828a179e6a53091496fddcee6e165e.

> And a third thing (sorry, should have thought of these all at once and sent fewer emails), can you add a CHANGES entry?

No problem, done in cc7272dfb5295631c5f252c87909ee29aa2f63ce.

sauclovian-g commented on intTests/test_type_errors/err028.log.good:
```diff
@@ -1,3 +1,12 @@
+    Found:    String
+
+    within "x" (err028.saw:7:5-7:6)
+
```
Can you also update the comments in the `*.saw` files for these? The message is no longer a parser message. It should probably say something like "Currently it trips on Block". There is a kind mismatch message, but it's still unreachable. (Except for `err033`, where it seems to have failed to engage, and that seems like a bug.)
